### PR TITLE
Add edmDumpClassVersion and checkDictionaryUpdate.py scripts to check class version and checksum updates

### LINKED ISSUE #947: Add utility to print class versions and checksums for automating setting `changes-dataformats` GitHub label
Motivation in https://github.com/cms-sw/cms-bot/issues/2245

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR is a work-in-progress prototype of a `edmDumpClassVersion` script to print class versions and checksums of all (non-transient) classes defined in a `classes_def.xml` file. The script would be used in automated assignment of `changes-dataformats` label discussed in https://github.com/cms-sw/cms-bot/issues/2245.

The command-line interface is similar to `edmCheckClassVersion` with the `-x / --xml_file` and `-l / --lib` parameters.

Resolves https://github.com/cms-sw/framework-team/issues/947

#### PR validation:

Tested the output on `DataFormats/Common/src/classes_def.xml`.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Potentially to be backported to 14_0_X.

### CODE REVIEW

**Review by makortel — COMMENTED** [review]
I think `FWCore/Utilities` is not the best package for these utilities (mostly because of the dependence on ROOT).

Maybe `FWCore/Reflection` would be better? (although I feel it would not be a perfect match either)

File: FWCore/Utilities/scripts/edmDumpClassVersion [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+
+import argparse
+
+import FWCore.Utilities.ClassesDefXmlUtils as ClassesDefUtils
+
+def main(args):
+    #Need to not have ROOT load .rootlogon.(C|py) since it can cause interference.
```
Comment (Dr15Jones) [inline FWCore/Utilities/scripts/edmDumpClassVersion:8]: As this part is also common, how about moving it to another .py file so it can be shared?

**Comment by makortel** [comment]
The present printout is along
```
edm::WrapperBase 4 3914746810
edm::EDProductGetter 11 2247759761
edm::RefCore 11 3425324092
edm::RefCoreWithIndex 11 2602044085
edm::ThinnedAssociation 10 1914950449
edm::Wrapper<edm::ThinnedAssociation> 4 2054679015
```

If we proceed by the bot prints the checksums of the baseline and of the baseline+PR for every PR test, this format might be sufficient (simple `diff` should be sufficient to tell if there are changes; in this case the output would probably be good to be sorted to avoid false positives from simple reordering of things in `classes_def.xml`).

If we proceed by having a storing the versions and checksums for all data format `classes_def.xml` files, the output would be better as something more machine-readable, e.g. JSON.

File: FWCore/Utilities/scripts/edmCheckClassVersion [hunk 2]
```diff
@@ -1,77 +1,7 @@
...
-                   ['short int',              'short'],
-                   ['long unsigned int',      'unsigned long'],
-                   ['unsigned long int',      'unsigned long'],
-                   ['long int',               'long'],
-                   ['std::string',            'std::basic_string<char>']] :
-            n_name = n_name.replace(e[0],e[1])
-        n_name = n_name.replace(' ','')
-        return n_name
+
+from FWCore.Utilities.ClassesDefXmlUtils import XmlParser
```
[the remark this answers is not in the document]
Comment (makortel) [reply FWCore/Utilities/scripts/edmCheckClassVersion:3]: I was trying to be a bit lazy, but I agree this doesn't really save much. I'll follow your suggestion.

File: FWCore/Utilities/scripts/edmDumpClassVersion (hunk 1 above)
Comment (makortel) [reply FWCore/Utilities/scripts/edmDumpClassVersion:8]: Agreed, I'll do that.

**Comment by makortel** [comment]
@smuzaffar What way do you think would make most sense to use the version+checksum information in the PR tests?

**Comment by makortel** [comment]
@smuzaffar Ok, sounds good. Do I understand correctly that a simple text format (like presently in this PR) would be sufficient?

What about packages that have multiple `classes_def.xml` files? I'd (slightly) prefer this script to operate on one `classes_def.xml` file at a time, so would the bot then e.g. concatenate+sort the output of all `classes_def.xml` files into one file per package?

**Comment by smuzaffar** [comment]
@makortel , for few `classes_def.xml` files `edmDumpClassVersion` failed with error like [a]. For packages which it failed are [b].

[a]
```
Traceback (most recent call last):
  File "/build/muz/cv/CMSSW_14_1_X_2024-07-11-2300/bin/el8_amd64_gcc12/edmDumpClassVersion", line 34, in <module>
    main(args)
  File "/build/muz/cv/CMSSW_14_1_X_2024-07-11-2300/bin/el8_amd64_gcc12/edmDumpClassVersion", line 22, in main
    (error, version, checksum) = ClassesDefUtils.checkClass(name, 0, {})
  File "/build/muz/cv/CMSSW_14_1_X_2024-07-11-2300/src/FWCore/Utilities/python/ClassesDefXmlUtils.py", line 95, in checkClass
    raise RuntimeError("failed to load dictionary for class '"+name+"'")
RuntimeError: failed to load dictionary for class 'std::pair<LocalVector,LocalVector>'
```

[b]
```
edmDumpClassVersion -l libAnalysisDataFormatsTrackInfo.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/AnalysisDataFormats/TrackInfo/src/classes_def.xml
edmDumpClassVersion -l libCondFormatsHcalObjects.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/CondFormats/HcalObjects/src/classes_def.xml
edmDumpClassVersion -l libDataFormatsEcalRecHitROCmAsync.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/EcalRecHit/src/alpaka/classes_rocm_def.xml
edmDumpClassVersion -l libDataFormatsHGCRecHit.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/HGCRecHit/src/classes_def.xml
edmDumpClassVersion -l libDataFormatsL1Trigger.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/L1Trigger/src/classes_def.xml
edmDumpClassVersion -l libDataFormatsTauReco.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/TauReco/src/classes_def_3.xml
edmDumpClassVersion -l libDataFormatsTrackSoAROCmAsync.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/TrackSoA/src/alpaka/classes_rocm_def.xml
edmDumpClassVersion -l libDataFormatsTrackingRecHitSoAROCmAsync.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/DataFormats/TrackingRecHitSoA/src/alpaka/classes_rocm_def.xml
edmDumpClassVersion -l libFireworksCore.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/Fireworks/Core/src/classes_def.xml
edmDumpClassVersion -l libOnlineDBEcalCondDB.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/OnlineDB/EcalCondDB/src/classes_def.xml
edmDumpClassVersion -l libSimGeneralTrackingAnalysis.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/SimGeneral/TrackingAnalysis/src/classes_def.xml
edmDumpClassVersion -l libSimTrackerTrackTriggerAssociation.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/SimTracker/TrackTriggerAssociation/src/classes_def.xml

```

**Comment by smuzaffar** [comment]
> @smuzaffar Ok, sounds good. Do I understand correctly that a simple text format (like presently in this PR) would be sufficient?

If I understand correctly, we do want to flag in case of any change in dataformat i.e. any new dict, removal of existing dict or change in checksum ... right? If yes then simple diff with this text format should be enough.
 
> What about packages that have multiple `classes_def.xml` files? I'd (slightly) prefer this script to operate on one `classes_def.xml` file at a time, so would the bot then e.g. concatenate+sort the output of all `classes_def.xml` files into one file per package?

yes, both will run single process per xml file but it will run `nproc` such process in parallel ( to speed up the processing). At the end it will concatenate+sort all the xmls for each package in its package file ( i.e. we will have one version-checksum file per package). Bot then can compare PR packages files with release packages file

**Comment by smuzaffar** [comment]
@makortel , I also see with classes with `-1` as version [a]. Is this expected ? 

[a]
```
> edmDumpClassVersion -l libCondFormatsHcalObjects.so -x /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-11-2300/src/CondFormats/HcalObjects/src/classes_def.xml
....
...
HcalCondObjectContainer<HcalLutMetadatum> -1 3488635253
HcalLutMetadata -1 1184932721
HcalLutMetadata::NonChannelData -1 1765044631
HcalDcsValue -1 3257269307

```

**Comment by makortel** [comment]
> > @smuzaffar Ok, sounds good. Do I understand correctly that a simple text format (like presently in this PR) would be sufficient?
> 
> If I understand correctly, we do want to flag in case of any change in dataformat i.e. any new dict, removal of existing dict or change in checksum ... right? 

Right, this is my understanding as well.

> If yes then simple diff with this text format should be enough.

Ok, good. I'll then update the PR with the earlier review suggestions.

> > What about packages that have multiple `classes_def.xml` files? I'd (slightly) prefer this script to operate on one `classes_def.xml` file at a time, so would the bot then e.g. concatenate+sort the output of all `classes_def.xml` files into one file per package?
> 
> yes, bot will run single process per xml file but it will run `nproc` such process in parallel ( to speed up the processing). At the end it will concatenate+sort all the xmls for each package in its package file ( i.e. we will have one version-checksum file per package). Bot then can compare PR packages files with release packages file

Sounds good.

**Comment by makortel** [comment]
> I also see some classes with `-1` as version [a]. Is this expected ?

I have no idea. I'll tag @pcanal in the hope we'd get an answer when he gets back.

**Comment by makortel** [comment]
> for few `classes_def.xml` files `edmDumpClassVersion` failed with error like [a]

Thanks, I'll take a look of these as well.

**Comment by smuzaffar** [comment]
> What about packages that have multiple classes_def.xml files?

yes bot will process all the xml files of the package. Also it will process any addition alpaka backend xml files
```
edmDumpClassVersion -l libDataFormatsStdDictionaries.so -x DataFormats/StdDictionaries/src/classes_def_vector.xml
edmDumpClassVersion -l libDataFormatsStdDictionaries.so -x DataFormats/StdDictionaries/src/classes_def_pair.xml
edmDumpClassVersion -l libDataFormatsStdDictionaries.so -x DataFormats/StdDictionaries/src/classes_def_others.xml
edmDumpClassVersion -l libDataFormatsStdDictionaries.so -x DataFormats/StdDictionaries/src/classes_def_map.xml
....
edmDumpClassVersion -l libDataFormatsBeamSpot.so -x DataFormats/BeamSpot/src/classes_def.xml
edmDumpClassVersion -l libDataFormatsBeamSpotCudaAsync.so -x DataFormats/BeamSpot/src/alpaka/classes_cuda_def.xml
edmDumpClassVersion -l libDataFormatsBeamSpotROCmAsync.so -x DataFormats/BeamSpot/src/alpaka/classes_rocm_def.xml
```

**Comment by davidlange6** [comment]
why would you expect a class version != -1 for these? 
https://github.com/cms-sw/cmssw/blob/c103a34c29b52f4d483433096101ac2e2403da4d/CondFormats/HcalObjects/src/classes_def.xml#L307

**Comment by makortel** [comment]
> why would you expect a class version != -1 for these?

Do you mean ROOT uses `-1` as the version for classes who don't have version set explicitly?

**Comment by davidlange6** [comment]
well I had thought so.. am I wrong? (maybe I am as I can't find something to confirm that:) )

**Comment by makortel** [comment]
> well I had thought so.. am I wrong? (maybe I am as I can't find something to confirm that:) )

I can believe it would be, I just don't know (and didn't quickly find anything about it, and @pcanal is away).

**Comment by makortel** [comment]
About the failures,

These
> ```
> AnalysisDataFormats/TrackInfo/src/classes_def.xml
> DataFormats/TauReco/src/classes_def_3.xml
> SimGeneral/TrackingAnalysis/src/classes_def.xml
> SimTracker/TrackTriggerAssociation/src/classes_def.xml
> ```
fail on loading a dictionary for `std::pair<X, Y>`. According to https://github.com/cms-sw/cmssw/issues/32444#issuecomment-742869118 the dictionaries for `std::pair<X, Y>` are not needed in our case, so I'll take a stab on removing all of those.

These
> ```
> CondFormats/HcalObjects/src/classes_def.xml
> Fireworks/Core/src/classes_def.xml
> OnlineDB/EcalCondDB/src/classes_def.xml
> ```
fail on base class meta information being unavailable. I'll dig deeper (I'm guessing the base class is missing from the `classes_def.xml`). `Fireworks` and `OnlineDB` are not data format packages, so we could just not run the script for those (in any case any differences in the checksums in those packages should not lead to `changes-dataformats` PR label)

These
> ```
> DataFormats/EcalRecHit/src/alpaka/classes_rocm_def.xml
> DataFormats/TrackSoA/src/alpaka/classes_rocm_def.xml
> DataFormats/TrackingRecHitSoA/src/alpaka/classes_rocm_def.xml
> ```
didn't reproduce on 14_1_0_pre5, so I need to try again on a newer IB. In any case the output should be empty (which makes me think if we'd want to use this tool to enforce the Alpaka `classes_{cuda,rocm}_def.xml` not defining any persistable data products?)

These
> ```
> DataFormats/HGCRecHit/src/classes_def.xml
> DataFormats/L1Trigger/src/classes_def.xml
> ```
complained about not being able to load a dictionary for `edm::Wrapper<X>`. I need to look deeper. Maybe some class is missing from the `classes_def.xml`?

**Comment by makortel** [comment]
> > ```
> > AnalysisDataFormats/TrackInfo/src/classes_def.xml
> > DataFormats/TauReco/src/classes_def_3.xml
> > SimGeneral/TrackingAnalysis/src/classes_def.xml
> > SimTracker/TrackTriggerAssociation/src/classes_def.xml
> > ```
> 
> 
> fail on loading a dictionary for `std::pair<X, Y>`. According to [#32444 (comment)](https://github.com/cms-sw/cmssw/issues/32444#issuecomment-742869118) the dictionaries for `std::pair<X, Y>` are not needed in our case, so I'll take a stab on removing all of those.

I have an attempt in https://github.com/cms-sw/cmssw/pull/45466

**Comment by makortel** [comment]
> > > ```
> > > AnalysisDataFormats/TrackInfo/src/classes_def.xml
> > > DataFormats/TauReco/src/classes_def_3.xml
> > > SimGeneral/TrackingAnalysis/src/classes_def.xml
> > > SimTracker/TrackTriggerAssociation/src/classes_def.xml
> > > ```
> > 
> > 
> > fail on loading a dictionary for `std::pair<X, Y>`. According to [#32444 (comment)](https://github.com/cms-sw/cmssw/issues/32444#issuecomment-742869118) the dictionaries for `std::pair<X, Y>` are not needed in our case, so I'll take a stab on removing all of those.
> 
> I have an attempt in #45466

Ok, it wasn't simple in the end. I'd need to dig into the framework's dictionary utilities to evolve the checks, and I'd need @pcanal's help for that (so this would be in some weeks).

**Comment by makortel** [comment]
An alternative "quick and dirty" path would be to ignore dictionary loading errors for now (could be good enough for the purposes of setting the `changes-dataformats` PR label), and work on those in longer term. This approach would also avoid having to backport any dictionary declaration fixes to 14_0_X.

**Comment by makortel** [comment]
It came to my mind that in principle we could use the planned mechanism of having a set of class versions+checksums of the base line compared to those from the baseline+PR to check and enforce a few other policies we have for the class versioning, such as "new class should start versioning from 3", "class update should increase the class version only by 1", "class template instantiations should not have class version set", and "non-dataformat packages may only define transient data products". For that kind of abilities having the information in  a more machine readable (e.h. JSON) format would be easier.

The (first use) case of `changes-dataformats` label could still be dealt with a diff between the (JSON) dumps. The other checks would, however, motivate another script, that would check if the class version+checksum of the PR are allowed given those of the baseline. Such a script could then also inform if dataformats were changed. 

(I certainly would not implement everything right a way, but I feel deciding if the cms-bot would run the same script on baseline and baseline+PR, or one script on baseline and another on baseline+PR would be worth of deciding now)

@smuzaffar What do you think?

**Comment by smuzaffar** [comment]
@makortel , extra class version checks you mentioned in https://github.com/cms-sw/cmssw/pull/45423#issuecomment-2231754855 sound great. 

About bot scripts, I was thinking to have one script to just dump class version+checksum information for both baseline and PRs. Then we can have another script to do the comparisons and report back issues e.g. if dataformats were changed or do not follow the properties you mentioned in https://github.com/cms-sw/cmssw/pull/45423#issuecomment-2231754855

**Comment by makortel** [comment]
> About bot scripts, I was thinking to have one script to just dump class version+checksum information for both baseline and PRs. Then we can have another script to do the comparisons and report back issues e.g. if dataformats were changed or do not follow the properties you mentioned in [#45423 (comment)](https://github.com/cms-sw/cmssw/pull/45423#issuecomment-2231754855)

Good idea, thanks. I'll craft the third script too in this PR, but only doing the "dataformats changed" check for now, leaving the rest for later. I was thinking the third script to
* return 0 and print nothing if there are no changes in persistable data formats
* return 1 and print e.g. "Changes in persistable data formats" when there are data format changes, and the policy checks succeed
* return 2 if policy checks fail and print some information why they failed.

**Comment by makortel** [comment]
I addressed previous comments by @Dr15Jones and myself, including changing the output format to JSON. By default the JSON document is printed to stdout, and I added `-o` / `--output` argument to write the JSON document directly to a file (the script may print warning messages, and therefore just piping the stdout to a file could produce invalid JSON document). I also made the script to ignore errors from dictionary loading for now (to make the backport to 14_0_X easier). I plan to look into those failures later on.

After discussing with @Dr15Jones I also moved the `edm{Check,Dump}ClassVersion` to `FWCore/Reflection`, and added some tests for them. While developing the tests, I noticed some of the errors from the scripts are not clear, so I improved some of those (and added tests).

I then added `checkDictionaryUpdate.py` script to `Utilities/ReleaseScripts`, that would then digest the JSON files from baseline and baseline+PR. Presently it only checks if the JSON documents are equal or not, but its interface is already built towards the policy checks (in the hope that a single update on the cms-bot would be sufficient). The possible exit codes are
* 0 for no changes
* 1 for a failure
* 2 for changes in dataformats, policy checks succeed
* 3 for policy checks fail

I noticed we have also `edmAddClassVersion` in `FWCore/Utilities` that contains similar code as in`edmCheckClassVersion`. I'll look into moving it to `FWCore/Reflection` and using the common components in a later PR.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@smuzaffar Please review :)

**Comment by makortel** [comment]
@smuzaffar One follow-up thought: would we want to run the policy checks in IBs (for the production flavor)? I'd expect we to have some number of policy violations presently, and tracking those _might_ be more convenient as an IB-level test.

**Comment by makortel** [comment]
Comparison differences are https://github.com/cms-sw/cmssw/issues/39803 and https://github.com/cms-sw/cmssw/issues/45505

**Comment by smuzaffar** [comment]
@makortel , so if I understand correct , bot should

- 0 for no changes :  every thing good and no need to add dataformats change label
- 1 for a failure: Failure in running script?? Should we mark PR test failure in this case?
- 2 for changes in dataformats, policy checks succeed:  Bot should add dataformats change label
- 3 for policy checks fail: What to do in this case? There might be already some policy failures... right, should we just add a warning in PR summary that dataformat policy check fail and convert that warning to error when master IB is clean of such errors?

> would we want to run the policy checks in IBs (for the production flavor)? 

yes, once this PR is merge then we can start running it for master IB production arch and report packages with policy checks failures

**Comment by makortel** [comment]
> * 0 for no changes :  every thing good and no need to add dataformats change label

:+1:

> * 1 for a failure: Failure in running script?? Should we mark PR test failure in this case?

I'd indeed mark the PR tests as failed in this case. It means the PR introduces some condition the script is not able to handle (for whatever reason), and we would need to be alerted to investigate. (Would it actually be possible for the bot to tag `@cms-sw/core-l2` in the PR test message in this case?) 

> * 2 for changes in dataformats, policy checks succeed:  Bot should add dataformats change label

:+1:

> * 3 for policy checks fail: What to do in this case? There might be already some policy failures... right, should we just add a warning in PR summary that dataformat policy check fail and convert that warning to error when master IB is clean of such errors?

I agree eventually a policy check failure should lead to marking PR tests as failed. I'm pretty sure we have some policy failures :) In the mean time (i.e. until we get IBs clean of such errors) I'm fine with reporting the policy failures as warnings.

> > would we want to run the policy checks in IBs (for the production flavor)?
> 
> yes, once this PR is merge then we can start running it for master IB production arch and report packages with policy checks failures

Ok. I'll add then a mode where the `checkDictionaryUpdate.py` can be run with `--baseline` only, and then I'd consider this PR complete (barring review feedback, of course).

File: FWCore/Reflection/test/BuildFile.xml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<library file="stubs/*.cc" name="FWCoreReflectionTestObjects">
+  <flags LCG_DICT_HEADER="stubs/classes.h"/>
+  <flags LCG_DICT_XML="stubs/classes_def.xml"/>
+  <use name="DataFormats/Common"/>
+</library>
+
+<test name="TestFWCoreReflectionCheckClassVersion" command="run_checkClassVersion.sh"/>
```
Comment (makortel) [inline FWCore/Reflection/test/BuildFile.xml:7]: @smuzaffar By the way, is there a away to denote a dependence from a test to the build rule that copies `scripts/*` to `$CMSSW_VERSION/bin/$SCRAM_ARCH`? Would e.g.
```xml
<test name="TestFWCoreReflectionCheckClassVersion" command="run_checkClassVersion.sh">
  <use name="FWCore/Reflection"/>
</test>
```
do that?

Comment (smuzaffar) [reply FWCore/Reflection/test/BuildFile.xml:7]: No @makortel , currently there is no way to provide such dependency

Comment (smuzaffar) [reply FWCore/Reflection/test/BuildFile.xml:7]: @makortel , it should be trivial to add such a support e.g. running the unit test rule can depend on `bin` and `script` rules of its dependencies

Comment (makortel) [reply FWCore/Reflection/test/BuildFile.xml:7]: Ok, thanks. I don't think the support would be crucial for this PR, but would be nice to have at some point.

**Comment by makortel** [comment]
The `checkDictionaryUpdate.py` can now be called with `--baseline` alone as well. While developing that, I realized using small values for the special exit codes was a bad idea (e.g. `argparse`'s `parse_args()` exits with `2` by default), so I renumbered the possible exit codes
 
* 0 for no changes : every thing good and no need to add dataformats change label
* 40 for changes in dataformats, policy checks succeed: Bot should add dataformats change label
* 41 for policy checks fail: start with a warning message, eventually (when we get IBs clean) mark PR tests as failed
* anything else is a failure in running the script: mark PR tests failed (and notify `@cms-sw/core-l2` if feasible)

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@smuzaffar The duplicate dictionary checker is (kind of rightfully) complaining about the tests
```
edm::Wrapper< edmtest::reflection::IntObject >
   ./FWCore/Reflection/test/stubs/classes_def.xml
   ./FWCore/Reflection/test/stubs/classes_def_ClassVersionMissingInClass.xml
   ./FWCore/Reflection/test/stubs/classes_def_ClassVersionMissingInVersion.xml
   ./FWCore/Reflection/test/stubs/classes_def_checksumMissingInVersion.xml
   ./FWCore/Reflection/test/stubs/classes_def_nameMissing.xml

edmtest::reflection::IntObject
   ./FWCore/Reflection/test/stubs/classes_def.xml
   ./FWCore/Reflection/test/stubs/classes_def_ClassVersionMissingInClass.xml
   ./FWCore/Reflection/test/stubs/classes_def_ClassVersionMissingInVersion.xml
   ./FWCore/Reflection/test/stubs/classes_def_checksumMissingInVersion.xml
```
Of these only `classes_def.xml` is used as part of build rules, and the others in test runner shell script. Therefore I could easily rename the `classes_def_*.xml` to e.g. `testclasses_def_*.xml`. @smuzaffar Would that be the way to go, or would you have any better suggestions?

**Comment by smuzaffar** [comment]
Duplicate dict script search for `*classes_def*.xml` (https://github.com/cms-sw/cmssw/blob/master/Utilities/ReleaseScripts/scripts/duplicateReflexLibrarySearch.py#L102) so renaming to something which does not have `classes_def` will work

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@smuzaffar Do you have any further comments?

**Comment by smuzaffar** [comment]
@makortel , all look good

**Comment by makortel** [comment]
+core

**Comment by mandrenguyen** [comment]
@makortel
Would you please resolve the open discussion threads, just to confirm that this is ready to merged?

**Comment by makortel** [comment]
> Would you please resolve the open discussion threads, just to confirm that this is ready to merged?

I'd actually prefer to leave them open. 3 of them are my own comments about why some things were done (and there is nothing to react), and 1 is a question/suggestion for @smuzaffar (for which there is nothing to react in this PR). I feel all of them are more useful to be left open than closed for anyone looking at this PR in the future.

**Comment by mandrenguyen** [comment]
> > Would you please resolve the open discussion threads, just to confirm that this is ready to merged?
> 
> I'd actually prefer to leave them open. 3 of them are my own comments about why some things were done (and there is nothing to react), and 1 is a question/suggestion for @smuzaffar (for which there is nothing to react in this PR). I feel all of them are more useful to be left open than closed for anyone looking at this PR in the future.

Fair enough.  But just to confirm, these are left open for the future, but the PR itself is ready to be merged?

**Comment by mandrenguyen** [comment]
+1

**Comment by makortel** [comment]
Coming back to the `classes_def.xml` files the script was unable to fully digest (i.e. continuation of https://github.com/cms-sw/cmssw/pull/45423#issuecomment-2226416814)

> ```
> AnalysisDataFormats/TrackInfo/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45564

> ```
> DataFormats/TauReco/src/classes_def_3.xml
> ```

I need to talk to @pcanal to understand better best practices with type aliases in `classes_def.xml` files.

> ```
> SimGeneral/TrackingAnalysis/src/classes_def.xml
> SimTracker/TrackTriggerAssociation/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45567


> ```
> CondFormats/HcalObjects/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45553

> ```
> Fireworks/Core/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45546

> ```
> OnlineDB/EcalCondDB/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45545

> ```
> DataFormats/EcalRecHit/src/alpaka/classes_rocm_def.xml
> DataFormats/TrackSoA/src/alpaka/classes_rocm_def.xml
> DataFormats/TrackingRecHitSoA/src/alpaka/classes_rocm_def.xml
> ```

I'll look into next week

> ```
> DataFormats/HGCRecHit/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45551

> ```
> DataFormats/L1Trigger/src/classes_def.xml
> ```

Fixed in https://github.com/cms-sw/cmssw/pull/45554

**Comment by smuzaffar** [comment]
>> yes, once this PR is merge then we can start running it for master IB production arch and report packages with policy checks failures

> Ok. I'll add then a mode where the checkDictionaryUpdate.py can be run with --baseline only, and then I'd consider this PR complete (barring review feedback, of course).

@makortel , class versions results for IB/baseline are now available ( see the link on Ib page : https://cmssdt.cern.ch/SDT/jenkins-artifacts//class_versions/CMSSW_14_1_X_2024-07-28-2300/el8_amd64_gcc12/class_versions.html ). There are few errors e.g. there are errors like
```
Parsing [CondFormats/CastorObjects/src/classes_def.xml](https://github.com/cms-sw/cmssw/blob/CMSSW_14_1_X_2024-07-28-2300/CondFormats/CastorObjects/src/classes_def.xml) failed: There is an element 'class' without 'name' attribute.
```

**Comment by makortel** [comment]
> class versions results for IB/baseline are now available ( see the link on Ib page : https://cmssdt.cern.ch/SDT/jenkins-artifacts//class_versions/CMSSW_14_1_X_2024-07-28-2300/el8_amd64_gcc12/class_versions.html ). There are few errors e.g. there are errors like

Thanks! I'll start taking a look. I see they all are about the `classes_def.xml` having `<class pattern="...">`. I vaguely recall a campaign to remove those some (many?) years ago.

**Pull request merged**